File: Acwing/First_Level_0_Basic_Algorithm/Second_DataStructure/ds_838_heap_sort/Main.java
package First_Level_0_Basic_Algorithm.Second_DataStructure.ds_838_heap_sort;

import java.util.Scanner;

public class Main {
    static int n , m , cnt;
    static final int N = 100010;
    static int [] h = new int[N];

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        m = sc.nextInt();
        for(int i = 1 ; i <= n ; ++i){
            h[i] = sc.nextInt();
        }
        cnt = n;
        for(int i = n / 2 ; i > 0 ; --i){
            down(i);
        }    
        while(m-- > 0){
            System.out.println(h[1] + " ");
            h[1] = h[cnt--];
            down(1);
        }
        sc.close();
    }

    public static void down(int u){
        int t = u;
        if(2 * u <= cnt && h[2 * u] < h[t]) t = 2 * u;
        if(2 * u + 1 <= cnt && h[2 * u + 1] < h[t]) t = 2 * u + 1;
        if(t != u){
            h[t] ^= h[u];
            h[u] ^= h[t];
            h[t] ^= h[u];
            down(t);
        }
    }
    
}

File: Acwing/First_Level_0_Basic_Algorithm/Second_DataStructure/ds_838_heap_sort/Main.cpp
#include<iostream>

using namespace std;
const int N = 100010;
int a[N] , cnt = 0;

void down(int x){
    int t = x;
    if(2 * x <= cnt && a[2 * x] < a[t]) t = 2 * x;
    if(2 * x + 1 <= cnt && a[2 * x + 1] < a[t]) t = 2 * x + 1;
    if(t != x) {
        swap(a[t] , a[x]);
        down(t);
    }
}

int main(){
    int n , m;
    cin >> n >> m;
    for(int i = 1 ; i <= n ; ++i){
        scanf("%d" , &a[i]);
    }
    cnt = n;
    for(int i = n / 2 ; i ; --i) down(i);
    
    while(m--){
       printf("%d " , a[1]);
       a[1] = a[cnt--];
       down(1);
    }
}